### DISCUSSION TITLE
bulk_insert insert first incomplete instances before inserting again

### DISCUSSION THREAD

**Discussion by thomashirtz** [sqlalchemy/sqlalchemy#5928] [Usage Questions] [answered]
**Describe the bug**
When I insert instances using bulk insert, first the instances are added without the relations, then added again. Which causes duplicates and incomplete rows

**Expected behavior**
The instances should only appear once inside the database.

**To Reproduce**
This is a very minimal version of my code, I kept the important pieces:
```
from sqlalchemy.orm import relationship
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Integer, String, ForeignKey, Float
from sqlalchemy.orm import sessionmaker
from sqlalchemy import create_engine
from contextlib import contextmanager
from sqlalchemy.exc import IntegrityError, InvalidRequestError

Base = declarative_base()


class Characteristic(Base):
    __tablename__ = "characteristic"
    id = Column('id', Integer, primary_key=True)
    name = Column('name', String, index=True)
    key = Column('key', String, index=True)
    measurements = relationship("Measurement", backref="characteristic")

    def __repr__(self):
        return f"<Characteristic(id='{self.id}', name='{self.name}', key='{self.key}')>"


class Measurement(Base):
    __tablename__ = "measurement"
    measurement_id = Column('measurement_id', Integer, primary_key=True)
    index = Column('index', Integer)
    value = Column('value', Float)
    simulation_id = Column('simulation_id', Integer, ForeignKey('simulation.id'))
    characteristic_id = Column('characteristic_id', Integer, ForeignKey('characteristic.id'))

    def __repr__(self):
        return f"<Measurement(id='{self.measurement_id}', simulation_id='{self.simulation_id}, characteristic_id='{self.characteristic_id}', index='{self.index}', value='{self.value}')>"


class Simulation(Base):
    __tablename__ = "simulation"
    id = Column('id', Integer, primary_key=True)
    measurements = relationship("Measurement", backref="simulation")

    def __repr__(self):
        return f"<Simulation(id='{self.id}')>"


@contextmanager
def session_scope(sessionmaker, **kwargs):
    """Provide a transactional scope around a series of operations."""
    session = sessionmaker(**kwargs)
    try:
        yield session
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()


def silent_insert(session, instance):
    try:
        session.add(instance)
        session.commit()
    except (IntegrityError,
            InvalidRequestError):
        print("silent_insert: Instance already present in the table")


def silent_bulk_insert(session, instances):
    try:
        session.bulk_save_objects(instances)
        session.commit()
    except (IntegrityError, InvalidRequestError):
        print("silent_bulk_insert: Instance already present in the table")


def find_instance(session, base: Base, column_to_value: dict):
    criterions = (getattr(base, column).like(value) for column, value in column_to_value.items())
    return session.query(base).filter(*criterions).first()


# Engine initialization
engine = create_engine(r'sqlite:///test.db', echo=True)
Session = sessionmaker(bind=engine)
Base.metadata.drop_all(bind=engine)

# Database Initialization
Base.metadata.create_all(bind=engine)
with session_scope(Session) as session:
    silent_insert(session, Simulation(id=1))
    silent_insert(session, Characteristic(name='drain', key='current'))

# Adding the data with bulk insert
with session_scope(Session) as session:
    instances = []
    simulation = find_instance(session, Simulation, {'id': 1})
    characteristic = find_instance(session, Characteristic, {'name': 'drain', 'key': 'current'})
    for i in range(5):
        instances.append(
            Measurement(**{'index': 1, 'value': 1, 'simulation': simulation, 'characteristic': characteristic}))

    silent_bulk_insert(session, instances)

# Checking the table
with session_scope(Session) as session:
    rows = session.query(Measurement).all()
    for row in rows:
        print(row)
```

**Error**
This is the output of the console:
```
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine ('drain', 'current', 1, 0)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine INSERT INTO measurement ("index", value) VALUES (?, ?)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine ((0, 0.0), (1, 2.0), (2, 4.0), (3, 6.0), (4, 8.0))
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine INSERT INTO measurement ("index", value, simulation_id, characteristic_id) VALUES (?, ?, ?, ?)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine (0, 0.0, 1, 1)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine INSERT INTO measurement ("index", value, simulation_id, characteristic_id) VALUES (?, ?, ?, ?)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine (1, 2.0, 1, 1)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine INSERT INTO measurement ("index", value, simulation_id, characteristic_id) VALUES (?, ?, ?, ?)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine (2, 4.0, 1, 1)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine INSERT INTO measurement ("index", value, simulation_id, characteristic_id) VALUES (?, ?, ?, ?)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine (3, 6.0, 1, 1)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine INSERT INTO measurement ("index", value, simulation_id, characteristic_id) VALUES (?, ?, ?, ?)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine (4, 8.0, 1, 1)
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine COMMIT
2021-02-07 22:21:46,621 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2021-02-07 22:21:46,621 INFO sqlalchemy.engine.base.Engine SELECT measurement.measurement_id AS measurement_measurement_id, measurement."index" AS measurement_index, measurement.value AS measurement_value, measurement.simulation_id AS measurement_simulation_id, measurement.characteristic_id AS measurement_characteristic_id 
FROM measurement
2021-02-07 22:21:46,621 INFO sqlalchemy.engine.base.Engine ()
<Measurement(id='1', simulation_id='None, characteristic_id='None', index='0', value='0.0')>
<Measurement(id='2', simulation_id='None, characteristic_id='None', index='1', value='2.0')>
<Measurement(id='3', simulation_id='None, characteristic_id='None', index='2', value='4.0')>
<Measurement(id='4', simulation_id='None, characteristic_id='None', index='3', value='6.0')>
<Measurement(id='5', simulation_id='None, characteristic_id='None', index='4', value='8.0')>
<Measurement(id='6', simulation_id='1, characteristic_id='1', index='0', value='0.0')>
<Measurement(id='7', simulation_id='1, characteristic_id='1', index='1', value='2.0')>
<Measurement(id='8', simulation_id='1, characteristic_id='1', index='2', value='4.0')>
<Measurement(id='9', simulation_id='1, characteristic_id='1', index='3', value='6.0')>
<Measurement(id='10', simulation_id='1, characteristic_id='1', index='4', value='8.0')>
2021-02-07 22:21:46,621 INFO sqlalchemy.engine.base.Engine COMMIT

Process finished with exit code 0
```
I guess the error come from this line ? 
```
2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine ((0, 0.0), (1, 2.0), (2, 4.0), (3, 6.0), (4, 8.0))
```

Basically I would only want that inside my measurement database:
```
<Measurement(id='1', simulation_id='1, characteristic_id='1', index='0', value='0.0')>
<Measurement(id='2', simulation_id='1, characteristic_id='1', index='1', value='2.0')>
<Measurement(id='3', simulation_id='1, characteristic_id='1', index='2', value='4.0')>
<Measurement(id='4', simulation_id='1, characteristic_id='1', index='3', value='6.0')>
<Measurement(id='5', simulation_id='1, characteristic_id='1', index='4', value='8.0')>
```
And not the other rows

**Versions.**
 - OS: Windows 10
 - Python: 3.8
 - SQLAlchemy: 1.3.20
 - SQLite 3.33.0

Thanks!

**Comment by zzzeek** [comment, marked as answer]
hi -

it looks like you are having the same issue as #5833, explanation is at https://github.com/sqlalchemy/sqlalchemy/issues/5833#issuecomment-758833193 .  

that is, your Measurement objects are being added to the Session and are part of the flush because you are associating them with "simulation" and "characteristic".    There is no reason to be setting these, as [bulk_insert will ignore them](https://docs.sqlalchemy.org/en/13/orm/persistence_techniques.html#bulk-operations-caveats) and you can see in your SQL output they are ignored (not in the column list or parameters):

    INSERT INTO measurement ("index", value) VALUES (?, ?)
    2021-02-07 22:21:46,611 INFO sqlalchemy.engine.base.Engine ((0, 0.0), (1, 2.0), (2, 4.0), (3, 6.0), (4, 8.0))

the INSERT statements that follow are the ones that occur as a result of the flush, because of this association.

The ways I would fix this include, in order of what I would do myself:

1. don't use bulk_insert(), it's not very useful and it's confusing, I should likely move to deprecate it due to the complicated boundaries it has.   The [big paragraph](https://docs.sqlalchemy.org/en/13/orm/persistence_techniques.html#bulk-operations) I added right at the top recently is intended to begin to tell people these methods are really not that useful.

2. when using bulk_insert, don't pass related objects to the target objects.  in this example you would need to send simulation_id and characteristic_id integer values directly.

3. set [cascade_backrefs=False](https://docs.sqlalchemy.org/en/13/orm/relationship_api.html#sqlalchemy.orm.relationship.params.cascade_backrefs) on Characteristic/Simulation to prevent the cascade from occurring (but you'd still need to do item 2).

**Reply by thomashirtz** [reply]
Thank you for your quick answer !

I did learn a lot from your message. Although I have a quick question, you advise me to not use bulk_insert, however, would it be also true if the number of measurements were significantly larger ? I need to insert ~1500 measurements (Each thread can have all of the 1500 measurement at the same moment, which is why I thought bulk seemed so convenient) per thread every 30 seconds on 4 to 32 threads. Doing that with a loop and a simple insert won't cause trouble ?

I have other bases, should I set the cascade backref to the other parents to False ? I tried to read the [doc](https://docs.sqlalchemy.org/en/13/orm/cascades.html#controlling-cascade-on-backrefs), but I am still a little bit confused about why you advise to set it at False in this specific use-case

Thanks again :)

**Reply by zzzeek** [reply]
> Thank you for your quick answer !
> 
> I did learn a lot from your message. Although I have a quick question, you advise me to not use bulk_insert, however, would it be also true if the number of measurements were significantly larger ? I need to insert ~1500 measurements (Each thread can have all of the 1500 measurement at the same moment, which is why I thought bulk seemed so convenient) per thread every 30 seconds on 4 to 32 threads. Doing that with a loop and a simple insert won't cause trouble ?

1500 is not that big of a number.  The overhead that the "bulk" system saves is on all the ORM-ish computation internally, such as taking relationships into account and such, as well as having to access newly generated primary keys, which is the big one that prevents a so-called "executemany' style run if the PKs are not available up front.

However there's a whole bunch of complicating factors for what the end result actually is, particularly if you're using PostgreSQL, as psycopg2's "executemany()" is still extremely slow unless you use some [flags](https://docs.sqlalchemy.org/en/13/dialects/postgresql.html#psycopg2-fast-execution-helpers) that are not the default in SQLAlchemy 1.3.     In SQLAlchemy 1.4, [we can now use PostgreSQL executemany in all cases, with all the fastest helpers](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#psycopg2-dialect-features-execute-values-with-returning-for-insert-statements-by-default) so if you were using psycopg2 you'd get as much as a 400% speedup using the vanilla ORM with SQLAlchemy 1.4, vs. wrestling with "bulk_insert" in SQLAlchemy 1.3.

Then the other thing with bulk_insert is that it still uses ORM objects which leads to all these confusing problems.  So I think people would be less confused, and they'd get faster performance anyway, if they just used table.insert() directly, where there is no chance of this kind of problem where the same ORM object is passed to bulk_insert() and also add'ed to the session, etc.   At least [bulk_insert_mappings()](https://docs.sqlalchemy.org/en/14/orm/contextual.html?highlight=bulk_insert_mappings#sqlalchemy.orm.scoping.scoped_session.bulk_insert_mappings) avoids using the ORM objects, which are themselves relatively expensive to create in Python vs. that of a dicionary alone.


> 
> I have other bases, should I set the cascade backref to the other parents to False ? I tried to read the [doc](https://docs.sqlalchemy.org/en/13/orm/cascades.html#controlling-cascade-on-backrefs), but I am still a little bit confused about why you advise to set it at False in this specific use-case

the "cascade backref" behavior is going away completely eventually because it is more often than not surprising and leads to this kind of problem.   In SQLAlchemy 1.4, if you use the Session() object with the future=True flag, then all cascade backrefs behavior turns off across the board.   so for your own applciation I would address this as a migration path you can take if and when you begin using the SQLAlchemy 1.4 series.

**Reply by thomashirtz** [reply]
@zzzeek I ran some benchmark and I got :
bulk_insert=False 35.73 seconds
bulk_insert=True 1.06 seconds
(per simulation)
(It runs on an old computer)

Currently the simulation takes ~100 seconds, so it means that it extends the simulation time by about 30+%. I really cannot afford this difference.. Is there much problems risk for production if I pass the ids directly as you mentioned (2), it seems to work fine right now, but I am not sure if in the long term problems may occurs (Maybe you witnessed such cases ?)

**Reply by zzzeek** [reply]
feel free to use bulk_insert() but if you want even more speed you would use connection.execute(table.insert()) directly.   that's my only point about bulk_insert() for all the confusion it causes.

**Reply by zzzeek** [reply]
also SQLAlhcemy 1.4 shoud be almost as fast as bulk_insert when using the ORM with psycopg2 because we've optimized the ORM for psycopg2's fast executemany mode, which is not in 1.3.
